src/ArrayHelper.php

```diff
@@ -428,20 +432,45 @@ public static function setValueByPath(array &$array, $path, $value, string $deli
 
     /**
      * @param array|float|int|string $path The path of where do you want to write a value to `$array`.
-     * The path can be described by a string when each key should be separated by delimiter.
+     * The path can be described by a string when each key should be separated by delimiter. If a path item contains
+     * delimiter, it can be escaped with "\" (backslash) or a custom delimiter can be used.
      * You can also describe the path as an array of keys.
-     * @param string $delimiter A separator, used to parse string $key for embedded object property retrieving. Defaults
+     * @param string $delimiter A separator, used to parse string key for embedded object property retrieving. Defaults
      * to "." (dot).
+     * @param bool $escapeDelimiter Whether to escape delimiter in the items of final array (in case of using string as
+     * an input) When `true`, "\" (backslashes) are removed - "\." will become "." for "." delimiter. Defaults to
+     * `false meaning backslashes will persist.
      *
      * @psalm-param ArrayPath $path
      *
      * @return array|float|int|string
      * @psalm-return ArrayKey
      */
-    private static function parsePath($path, string $delimiter)
+    public static function parsePath($path, string $delimiter = '.', bool $escapeDelimiter = false)
     {
         if (is_string($path)) {
-            return explode($delimiter, $path);
+            if (strlen($delimiter) !== 1) {
+                throw new InvalidArgumentException('Only 1 character is allowed for delimiter.');
+            }
+
+            if (($path[0] ?? '') === $delimiter) {
+                throw new InvalidArgumentException('Delimiter can\'t be at the very beginning.');
+            }
+
+            if (substr($path, -1) === $delimiter && substr($path, -2) !== '\\' . $delimiter) {
+                throw new InvalidArgumentException('Delimiter can\'t be at the very end.');
+            }
+
+            $pattern = sprintf('/(?:[^\%s\\\\]|\\\\.)+/', $delimiter);
+            preg_match_all($pattern, $path, $matches);
+
+            if ($escapeDelimiter === true) {
```

Comment:
Have a case that `escapeDelimiter` should be is true? May be we can remove this argument?

Reply:
Here - https://github.com/yiisoft/validator/blob/d5bc83d5c973e14840286e60a09813a360cf6849/src/Rule/Nested.php#L121. Added exactly for this case.

Reply:
Ok, then sugget rename him to `$keepDelimiterEscaping`. What you think?

Reply:
Yes, it's a good alternative because it highlights that escaping is not performed again, rather additional unescaping is done (this was mentioned in initial name). Subjective, but "preserve" is slightly better in my opinion.